File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>, 
        mut list2: Option<Box<ListNode>>
    ) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;

        while let (Some(node1), Some(node2)) = (&list1, &list2) {
            if node1.val <= node2.val {
                tail.next = list1.take();
                tail = tail.next.as_mut().unwrap();
                list1 = tail.next.take();
            } else {
                tail.next = list2.take();
                tail = tail.next.as_mut().unwrap();
                list2 = tail.next.take();
            }
            
        }

        tail.next = list1.or(list2);
        dummy.next
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverse(self, head):
        curr = head
        prev = None

        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        return prev

    def reverseKGroup(self, head, k):
        dummy = ListNode(0)
        tail = dummy
        group_head = head

        while group_head:
            kth = group_head

            for _ in range(k - 1):
                if kth is None:
                    break
                kth = kth.next

            if kth is None:
                tail.next = group_head
                break

            group_next = kth.next
            kth.next = None

            new_head = self.reverse(group_head)
            tail.next = new_head

            tail = group_head
            group_head = group_next

        return dummy.next